### PR TITLE
[tmva][sofie] Account for dilation in Conv2D same padding calculation

### LINKED ISSUE #21684: [tmva][sofie] Keras parser: Conv2D padding='same' ignores dilation_rate, producing wrong padding
### Check duplicate issues.

- [ ] Checked for duplicates

### Description

In `bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_sofie/_parser/_keras/layers/conv.py`, the `padding='same'` calculation does not account for `dilation_rate`. The effective kernel size for a dilated convolution is `dilation * (kernel - 1) + 1`, but the code uses the raw kernel size directly:

```
# lines 55-56 — dilation_rate is extracted at line 36 but never used here
padding_height = max((outputHeight - 1) * fAttrStrides[0] + fAttrKernelShape[0] - inputHeight, 0)
padding_width  = max((outputWidth  - 1) * fAttrStrides[1] + fAttrKernelShape[1] - inputWidth,  0)
```

For `dilation=2, kernel=3, input=8, stride=1`:
- SOFIE computes total padding = 2
- Correct total padding = 4 (verified against Keras output shape)
- With half the required padding, the generated inference code operates on tensors with wrong spatial dimensions

### Reproducer

Step 1 — verify Keras ground truth: `padding='same'` with `dilation_rate=2` should preserve spatial dimensions:

```python
import keras
model = keras.Sequential([
    keras.Input(shape=(8, 8, 1)),
    keras.layers.Conv2D(1, kernel_size=3, padding='same', dilation_rate=2)
])
print('input shape: ', model.input_shape)   # (None, 8, 8, 1)
print('output shape:', model.output_shape)  # (None, 8, 8, 1)  <- same padding must preserve this
```

Step 2 — show what SOFIE actually computes vs the correct value:

```python
import math

inputHeight    = 8
fAttrStrides   = [1, 1]
fAttrKernelShape = [3, 3]
fAttrDilations = [2, 2]

outputHeight = math.ceil(inputHeight / fAttrStrides[0])

buggy   = max((outputHeight - 1) * fAttrStrides[0] + fAttrKernelShape[0] - inputHeight, 0)
eff_k   = fAttrDilations[0] * (fAttrKernelShape[0] - 1) + 1
correct = max((outputHeight - 1) * fAttrStrides[0] + eff_k - inputHeight, 0)

print(f'SOFIE computes:  {buggy}')   # 2
print(f'correct padding: {correct}') # 4
```

Image

### ROOT version

6.39.01 (built from source)

### Installation method

Built from source with `-Dtmva-sofie=ON -Dtmva-pymva=ON -Dbuiltin_protobuf=ON`

### Operating system

Linux (Ubuntu 22.04)

### Additional context

Only affects Conv2D with `padding='same'` and `dilation_rate > 1`. The padding='valid' path is unaffected since it does not compute explicit padding values.

### PR DESCRIPTION
**Pull request by harz05**
## Changes

The `padding='same'` calculation in the Keras Conv2D parser was using the raw kernel size instead of the effective kernel size, which for dilated convolutions is `dilation * (kernel - 1) + 1`. `fAttrDilations` was extracted from the layer attributes but never factored into the padding formula.

```python
# before
padding_height = max((outputHeight - 1) * strides[0] + kernel_shape[0] - inputHeight, 0)

# after
effective_kH = dilations[0] * (kernel_shape[0] - 1) + 1
padding_height = max((outputHeight - 1) * strides[0] + effective_kH - inputHeight, 0)
```

For `dilation=2, kernel=3, input=8, stride=1` the old formula produced padding=2 while the correct value is 4, causing the generated inference code to operate on tensors with wrong spatial dimensions. With `dilation=1` the effective kernel size reduces to the raw kernel size, so existing models are unaffected.

## Checklist

- [x] tested changes locally

This PR fixes #21684

### CODE REVIEW

**Review by sanjibansg — DISMISSED** [review]
Hi @harz05,
Thanks for this PR. Can you add a test so that we can verify this patch?

**Comment by harz05** [comment]
Hello @sanjibansg, I tried adding a numerical test with Conv2D(kernel=3, padding='same', dilation_rate=2) on 8×8 input and on testing locally I found that the output shape was correct (8x8x4=256 vs 144 before the fix), confirming the padding calculation works.
However the numerical values don't match Keras. The reason I could figure out is that ROperator_Conv::ShapeInference in `ROperator_Conv.hxx` overwrites `fAttrKernelShape` with the dilated kernel size and then `Im2col` applies dilation again. This is a seperate double dilation bug and a full numerical test might fail due to this and not due to the padding fix.
Should I raise a follow up issue first?

**Comment by sanjibansg** [comment]
> Hello @sanjibansg, I tried adding a numerical test with Conv2D(kernel=3, padding='same', dilation_rate=2) on 8×8 input and on testing locally I found that the output shape was correct (8x8x4=256 vs 144 before the fix), confirming the padding calculation works. However the numerical values don't match Keras. The reason I could figure out is that ROperator_Conv::ShapeInference in `ROperator_Conv.hxx` overwrites `fAttrKernelShape` with the dilated kernel size and then `Im2col` applies dilation again. This is a seperate double dilation bug and a full numerical test might fail due to this and not due to the padding fix. Should I raise a follow up issue first?

Hi @harz05,
Could that be fixed within this PR itself, that should be fine and we can expand its scope.

**Comment by harz05** [comment]
hello @sanjibansg
i've changed `Im2col` calls in `ROperator_Conv.hxx` to pass dilation as 1, also added a dilated Conv2d test and corresponding model generator. Pls review it once.
Thank you.

**Comment by harz05** [comment]
There was a clang and ruff formatting issue during the run, I fixed that. CI was also getting Windows x86 failure but i'm not really sure what's the issue in that.
pls review

**Comment by harz05** [comment]
hello @sanjibansg ,sorry for the noisy commit history, I was trying to fix the ruff formatting issue and accidentally did a rebase that pulled in a ton of upstream master commits into the branch. I've force pushed a clean branch with just the 4 PR commits now
ruff format is fixed as flagged by CI
all other tests were passing according to the CI run

**Review by lmoneta — APPROVED** [review]
LGTM!
Can you please include the test not as a Python tests in the https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/test/sofie_keras_parser.py ? 

Or adding an ONNX file and integrate in https://github.com/root-project/root/blob/master/tmva/sofie/test/TestCustomModelsFromONNX.cxx

**Comment by harz05** [comment]
Hello @lmoneta , i've added Conv2D_padding_same_dilation (kernel=3, dilation_rate=2, 8x8x3 input) to `sofie_keras_parser.py` with generators for both sequential and functional model types.
pls review & verify.

**Review by guitargeek — APPROVED** [review]
LGTM, thanks for adding the fix and the test! I have rebased the PR on master, fixed the conflicts, and I'll merge it once the CI passes.

**Comment by harz05** [comment]
thanks for the merge
